File: DMC-E系列控制卡例程/VB.net/例6_下载配置文件/下载配置文件/下载配置文件/Form1.vb
Imports System.Threading
Imports System.IO
Imports System.Text


Public Class Form1
    Public g_sCardID As Short
    Private Sub Button4_Click(ByVal sender As System.Object, ByVal e As System.EventArgs)
        RichTextBox1.Text = ""
        RichTextBox1.AppendText("请勿操作，总线卡软件（热）复位进行中……")
        Btn_ClodReset.Enabled = False
        Btn_HoldReset.Enabled = False
        Dim dt_start As DateTime
        Dim sRtn As Short
        Dim usErr As UShort
        dt_start = DateTime.Now
        sRtn = dmc_soft_reset(g_sCardID)
        If sRtn <> 0 Then
            MessageBox.Show("dmc_soft_reset == " + sRtn.ToString())
            Return
        End If
        While True
            If (DateTime.Now - dt_start).TotalMilliseconds >= 10000.0 Then '设置超时退出，防止死循环
                MessageBox.Show("总线复位失败!")
                Btn_ClodReset.Enabled = True
                Btn_HoldReset.Enabled = True
                Btn_HoldReset.Focus()
                Return
            End If

            sRtn = nmc_get_errcode(g_sCardID, 2, usErr)
            If sRtn = 0 Then
                If usErr = 0 Then '总线复位正常完成
                    Exit While
                End If
            Else
                MessageBox.Show("nmc_get_errcode == " + sRtn.ToString())
                Btn_ClodReset.Enabled = True
                Btn_HoldReset.Enabled = True
                Btn_HoldReset.Focus()
                Return
            End If
            Application.DoEvents()
        End While
        RichTextBox1.AppendText("总线卡软件（热）复位完成,请确认总线状态")
        Btn_ClodReset.Enabled = True
        Btn_HoldReset.Enabled = True
        Btn_HoldReset.Focus()
    End Sub

    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Dim sNum As Short
        Dim arr_nCardtypes(8) As UInteger
        Dim arr_usCardids(8) As UShort
        sNum = dmc_board_init() '获取卡数量
        If (sNum <= 0) Or (sNum > 8) Then             '正常的卡数在1- 8之间
            MsgBox("初始化LTDMC卡失败！", vbOKOnly, "出错")
        Else
            MsgBox("初始化LTDMC卡成功！", vbOKOnly, "OK")
        End If
        dmc_get_CardInfList(sNum, arr_nCardtypes, arr_usCardids) '
        g_sCardID = arr_usCardids(1) '
        Timer1.Start() '
    End Sub

    Private Sub Timer1_Tick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Timer1.Tick
        Dim errcode As Int32
        nmc_get_errcode(g_sCardID, 2, errcode)
        If (errcode = 0) Then
            TextBox1.Text = "EtherCAT总线正常"
            TextBox1.BackColor = Color.Green
        Else
            TextBox1.Text = "EtherCAT总线出错"
            TextBox1.BackColor = Color.Red
        End If
    End Sub

  

    Private Sub Form1_FormClosing(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles MyBase.FormClosing
        Timer1.Stop()
    End Sub


    Private Sub Btn_DownINI_Click(sender As System.Object, e As System.EventArgs) Handles Btn_DownINI.Click
        Dim file2 = New OpenFileDialog()

        file2.Filter = "ini|*.ini"
        Dim sRtn = file2.ShowDialog()
        If sRtn = DialogResult.OK Then
            Dim str As String
            str = File.ReadAllText(file2.FileName)
            Dim buffer1() As Byte = Encoding.UTF8.GetBytes(str)
            Dim fileincontrol() As Byte = Encoding.UTF8.GetBytes("")
            Dim filetype2 As UShort = 201
            Dim ret As Short = dmc_download_memfile(g_sCardID, buffer1, buffer1.Length, fileincontrol, filetype2)

            If (ret = 0) Then
                MessageBox.Show("下载ENI配置文件成功！")
            End If

        End If
    End Sub

    Private Sub Btn_DownENI_Click(sender As System.Object, e As System.EventArgs) Handles Btn_DownENI.Click
        Dim file2 = New OpenFileDialog()

        file2.Filter = "eni|*.eni"
        Dim sRtn = file2.ShowDialog()
        If sRtn = DialogResult.OK Then
            Dim str As String
            str = File.ReadAllText(file2.FileName)
            Dim buffer1() As Byte = Encoding.UTF8.GetBytes(str)
            Dim fileincontrol() As Byte = Encoding.UTF8.GetBytes("")
            Dim filetype2 As UShort = 200
            Dim ret As Short = dmc_download_memfile(g_sCardID, buffer1, buffer1.Length, fileincontrol, filetype2)

            If (ret = 0) Then
                MessageBox.Show("下载ENI配置文件成功！")
            End If

        End If
    End Sub

    Private Sub Btn_ClodReset_Click(sender As System.Object, e As System.EventArgs) Handles Btn_ClodReset.Click
        RichTextBox1.Text = ""
        RichTextBox1.AppendText("请勿操作，总线卡硬件（冷）复位进行中……")
        'Button3.Enabled = False
        'Button4.Enabled = False
        dmc_board_reset()
        dmc_board_close()
        For i As Integer = 0 To 15 Step 1 '总线卡硬件复位耗时15s左右
            Application.DoEvents()
            Thread.Sleep(1000)
        Next
        dmc_board_init()
        RichTextBox1.AppendText("总线卡硬件（冷）复位完成,请确认总线状态")
        'Button3.Enabled = True
        'Button4.Enabled = True
        'Button3.Focus()
    End Sub

    Private Sub Btn_HoldReset_Click(sender As System.Object, e As System.EventArgs) Handles Btn_HoldReset.Click
        RichTextBox1.Text = ""
        RichTextBox1.AppendText("请勿操作，总线卡软件（热）复位进行中……")
        Btn_ClodReset.Enabled = False
        Btn_HoldReset.Enabled = False
        Dim dt_start As DateTime
        Dim sRtn As Short
        Dim usErr As UShort
        dt_start = DateTime.Now
        sRtn = dmc_soft_reset(g_sCardID)
        If sRtn <> 0 Then
            MessageBox.Show("dmc_soft_reset == " + sRtn.ToString())
            Return
        End If
        While True
            If (DateTime.Now - dt_start).TotalMilliseconds >= 10000.0 Then '设置超时退出，防止死循环
                MessageBox.Show("总线复位失败!")
                Btn_ClodReset.Enabled = True
                Btn_HoldReset.Enabled = True
                Btn_HoldReset.Focus()
                Return
            End If

            sRtn = nmc_get_errcode(g_sCardID, 2, usErr)
            If sRtn = 0 Then
                If usErr = 0 Then '总线复位正常完成
                    Exit While
                End If
            Else
                MessageBox.Show("nmc_get_errcode == " + sRtn.ToString())
                Btn_ClodReset.Enabled = True
                Btn_HoldReset.Enabled = True
                Btn_HoldReset.Focus()
                Return
            End If
            Application.DoEvents()
        End While
        RichTextBox1.AppendText("总线卡软件（热）复位完成,请确认总线状态")
        Btn_ClodReset.Enabled = True
        Btn_HoldReset.Enabled = True
        Btn_HoldReset.Focus()
    End Sub

    Private Sub Btn_InitialReset_Click(sender As System.Object, e As System.EventArgs) Handles Btn_InitialReset.Click
        Dim sRtn As Short
        RichTextBox1.Text = ""
        RichTextBox1.AppendText("请勿操作，总线卡初始值复位进行中……")
        '‘Button3.Enabled = False
        'Button4.Enabled = False
        '调用该函数只是删除了文件信息，此时需要，再调用一次冷复位或者热复位重启总线才可以完全清除系统的临时缓存信息。
        sRtn = dmc_original_reset(g_sCardID)
        If sRtn <> 0 Then
            MessageBox.Show("dmc_original_reset == " + sRtn.ToString())
            Return
        End If
        For i As Integer = 0 To 2 Step 1 '总线卡初始值复位耗时2s左右
            Application.DoEvents()
            Thread.Sleep(1000)
        Next
        sRtn = dmc_soft_reset(g_sCardID)
        If sRtn <> 0 Then
            MessageBox.Show("dmc_soft_reset == " + sRtn.ToString())
            Return
        End If
        'dmc_board_close()
        For i As Integer = 0 To 15 Step 1 '总线卡初始值复位耗时15s左右
            Application.DoEvents()
            Thread.Sleep(1000)
        Next
        'dmc_board_init()
        RichTextBox1.AppendText("总线卡初始值复位完成,请确认总线状态")
        ' Button3.Enabled = True
        ' Button4.Enabled = True
        ' Button4.Focus()
    End Sub

    Private Sub Btn_AxisEnable_Click(sender As System.Object, e As System.EventArgs) Handles Btn_AxisEnable.Click
        nmc_set_axis_enable(g_sCardID, 255)
    End Sub

    Private Sub Btn_AxisDisable_Click(sender As System.Object, e As System.EventArgs) Handles Btn_AxisDisable.Click
        nmc_set_axis_disable(g_sCardID, 255)
    End Sub

    Private Sub Btn_Close_Click(sender As System.Object, e As System.EventArgs) Handles Btn_Close.Click
        Me.Close()
    End Sub

    Private Sub Button1_Click(sender As System.Object, e As System.EventArgs) Handles Button1.Click
        Dim file1 = New OpenFileDialog()
        file1.Filter = "s6x|*.s6x"

        Dim sRtn = file1.ShowDialog()
        If sRtn = DialogResult.OK Then
            Dim str As String
            str = file1.FileName
            Dim ret As Short = dmc_download_firmware(g_sCardID, str)

            If (ret = 0) Then
                MessageBox.Show("下载固件文件成功！")
            End If

        End If
    End Sub
End Class

File: DMC-E系列控制卡例程/c#/例6_下载EtherCAT配置文件/下载EtherCAT配置文件/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.IO;
using csLTDMC;
using System.Threading;

namespace 下载EtherCAT配置文件
{
    public partial class Form1 : Form
    {
        private ushort _CardNo = 0;
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            OpenFileDialog f = new OpenFileDialog();
            f.Filter = "ini|*.ini";
            DialogResult res = f.ShowDialog();
            if (res == DialogResult.OK)
            {
                FileStream fs = File.Open(f.FileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
                StreamReader sr = new StreamReader(fs);
                string str = sr.ReadToEnd();
                sr.Close();
                fs.Close();
                byte[] buffer = Encoding.UTF8.GetBytes(str);
                byte[] fileincontrol = Encoding.UTF8.GetBytes("");
                ushort filetype = 201;
                short ret = -1;
                ret = LTDMC.nmc_set_cycletime(_CardNo, 2, 1000);      //设置总线周期
                ret = LTDMC.dmc_download_memfile(_CardNo, buffer, (uint)buffer.Length, fileincontrol, filetype);     //下载INI文件
                if (ret == 0)
                {
                    MessageBox.Show("下载成功！");
                }
            }
        }

        private void timer1_Tick(object sender, EventArgs e)
        {
            ushort errcode = 0;
            LTDMC.nmc_get_errcode(_CardNo, 2, ref errcode);
            if (errcode == 0)
            {
                textBox5.Text = "总线正常";
                textBox5.BackColor = Color.Green;
            }
            else
            {
                textBox5.Text = "总线错误";
                textBox5.BackColor = Color.Red;
            }
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            short num = -1;
            num = LTDMC.dmc_board_init();
            if (num <= 0 || num >= 8)
            {
                MessageBox.Show("初始化控制卡失败！", "出错");
            }
            ushort _num = 0;
            ushort[] cardids = new ushort[8];
            uint[] cardtypes = new uint[8];
            short res = LTDMC.dmc_get_CardInfList(ref _num, cardtypes, cardids);
            if (res != 0)
            {
                MessageBox.Show("获取控制卡信息失败！");
            }
            _CardNo = cardids[0];
            timer1.Start();
        }

        private void button3_Click(object sender, EventArgs e)
        {
            LTDMC.nmc_set_axis_enable(_CardNo, 255);
        }

        private void button4_Click(object sender, EventArgs e)
        {
            LTDMC.nmc_set_axis_disable(_CardNo, 255);
        }

        private void Form1_FormClosed(object sender, FormClosedEventArgs e)
        {
            LTDMC.dmc_board_close();
        }

        private void button6_Click(object sender, EventArgs e)
        {
            OpenFileDialog f = new OpenFileDialog();
            f.Filter = "eni|*.eni";
            DialogResult res = f.ShowDialog();
            if (res == DialogResult.OK)
            {
                FileStream fs = File.Open(f.FileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
                StreamReader sr = new StreamReader(fs);
                string str = sr.ReadToEnd();
                sr.Close();
                fs.Close();
                byte[] buffer = Encoding.UTF8.GetBytes(str);
                byte[] fileincontrol = Encoding.UTF8.GetBytes("");
                ushort filetype = 200;
                short ret = -1;
                ret = LTDMC.dmc_download_memfile(_CardNo, buffer, (uint)buffer.Length, fileincontrol, filetype);     //下载ENI文件
                if (ret == 0)
                {
                    MessageBox.Show("下载成功！");
                }
            }
        }

        private void button7_Click(object sender, EventArgs e)
        {
            richTextBox1.Text = "";
            richTextBox1.AppendText("请勿操作，总线卡软件复位进行中……" + "\n");
            button1.Enabled = false;
            button2.Enabled = false;
            button3.Enabled = false;
            button4.Enabled = false;
            button5.Enabled = false;
            button6.Enabled = false;
            button7.Enabled = false;
            LTDMC.dmc_soft_reset(_CardNo);
            LTDMC.dmc_board_close();

            for (int i = 0; i < 15; i++)//总线卡软件复位耗时15s左右
            {
                Application.DoEvents();
                Thread.Sleep(1000);
            }

            LTDMC.dmc_board_init();
            richTextBox1.AppendText("总线卡软件复位完成,请确认总线状态");
            button1.Enabled = true;
            button2.Enabled = true;
            button3.Enabled = true;
            button4.Enabled = true;
            button5.Enabled = true;
            button6.Enabled = true;
            button7.Enabled = true;
        }

        private void button2_Click(object sender, EventArgs e)
        {
            richTextBox1.Text = "";
            richTextBox1.AppendText("请勿操作，总线卡硬件复位进行中……" + "\n");
            button1.Enabled = false;
            button2.Enabled = false;
            button3.Enabled = false;
            button4.Enabled = false;
            button5.Enabled = false;
            button6.Enabled = false;
            button7.Enabled = false;
            LTDMC.dmc_board_reset();
            LTDMC.dmc_board_close();

            for (int i = 0; i < 15; i++)//总线卡硬件复位耗时15s左右
            {
                Application.DoEvents();
                Thread.Sleep(1000);
            }

            LTDMC.dmc_board_init();
            richTextBox1.AppendText("总线卡硬件复位完成,请确认总线状态");
            button1.Enabled = true;
            button2.Enabled = true;
            button3.Enabled = true;
            button4.Enabled = true;
            button5.Enabled = true;
            button6.Enabled = true;
            button7.Enabled = true;
        }

        private void button5_Click(object sender, EventArgs e)
        {
            richTextBox1.Text = "";
            richTextBox1.AppendText("请勿操作，总线卡初始值复位进行中……" + "\n");
            button1.Enabled = false;
            button2.Enabled = false;
            button3.Enabled = false;
            button4.Enabled = false;
            button5.Enabled = false;
            button6.Enabled = false;
            button7.Enabled = false;
            LTDMC.dmc_original_reset(_CardNo);
            LTDMC.dmc_board_close();

            for (int i = 0; i < 15; i++)//总线卡硬件复位耗时15s左右
            {
                Application.DoEvents();
                Thread.Sleep(1000);
            }

            LTDMC.dmc_board_init();
            richTextBox1.AppendText("总线卡初始值复位完成,请确认总线状态");
            button1.Enabled = true;
            button2.Enabled = true;
            button3.Enabled = true;
            button4.Enabled = true;
            button5.Enabled = true;
            button6.Enabled = true;
            button7.Enabled = true;
        }
    }
}
